docs/specs/merkle_tree.md

```diff
@@ -0,0 +1,39 @@
+- read hash
+- read data
+- write data
+- retrieve merkle path
+
+When accessing information of the merkle tree, a merkle location is provided. A merkle location is a length encoded bitstring. Each bit represents a binary direction to take when traversing the merkle tree, starting from the first bit. If the bit is 0, traverse left, and if the bit is 1, traverse right.
+
+Reads are separated by hash and data as the return values are different. Both reads and writes are performed as bulk operations, meaning that the code that calls into these is responsible for batching as many reads and writes as it can at once. Batching of reads and writes is an important optimization for both the persistent database, and for the merkle tree masks. Writing data has the side effect of recalculating the preceeding hashes of that data.
+
+Retrieving a merkle path involves traversing a given data merkle location and returning the sibling of each node in the traversal.
+
+## Maskable Interface
+
+A maskable merkle tree is a superset of the base merkle tree interface which allows masks to be derived from the structure.
+
+There are also masking operations:
+- register mask
+- unregister mask
+
+Registering a mask creates a new mask as a child of the merkle tree. Whenever the maskable merkle tree receives a write request, it will notify all of its children with the data modified as part of the write. More details on this are in the "Mask" section.
+
+## Mask
+
+A merkle tree mask is a sparse collection of merkle tree nodes which are different from its parent. Whenever a read occurs on a mask, the mask will first attempt to retrieve the node from its internal sparse collection, and if it does not exist there, will forward the read request to its parent. Writing to a merkle tree mask will only update its internal sparse collection.
+
+The merkle tree mask implements a superset of the base merkle tree interface. The unique operations of a mask are:
+- commit
+- invalidate cache
+
+Committing a merkle tree mask will cause it to bulk write all of its changes to its parent, and then clear all of its internal state. This is useful for use masks as temporary views into a set of changes that you may or may not want to apply back to another merkle tree. For instance, you may want to apply transactions to a mask first and validate the merkle path of the new tree before you update the underlying merkle tree.
+
+Every mask provides a hook for invalidating its cache. The hook is intended to be called whenever the mask's parent is updated. This hook takes in a set of nodes that have been modified in that parent merkle tree. The invalidation rule is that if any of the nodes received by the hook are the same as the corresponding node stored internal to the mask, then that information is deleted from the mask. This works because if the parent merkle tree's node is the same as the mask's node, then reading from the mask's cache is equivalent to forwarding the read to the parent.
```

Comment:
Maybe "invalidate" is the wrong word here; it suggests the mask is corrupt or something. Maybe "update" instead?

Currently, the mask uses `Account.equal` to compare the parent node with its own. Probably it should check just equality of the public key to see whether it can delete a node, yes?

Also, if the accounts differ, the mask needs to relocate its own node, so there's no conflict. Should that be mentioned here (I think not, but ...).